stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+                match http_response_preamble.content_length {
+                    Some(len) => Some(len as usize),
+                    None => None,
+                }
+            }
+        }
+    }
+
+    /// Read the next HTTP preamble (be it a request or a response), and return the preamble and
+    /// the number of bytes consumed while reading it.
+    fn read_preamble(&mut self, buf: &[u8]) -> Result<(StacksHttpPreamble, usize), NetError> {
+        // does this contain end-of-headers marker, including the last four bytes of preamble we
+        // saw?
+        if self.body_start.is_none() {
+            for i in 0..=buf.len() {
+                let window = self.body_start_search_window(i, buf);
+                if window == [13, 10, 13, 10] {
+                    self.body_start = Some(self.num_preamble_bytes + i);
+                }
+            }
+        }
+        if self.body_start.is_none() {
+            // haven't found the body yet, so update `last_four_preamble_bytes`
+            // and report underflow
+            let len = buf.len();
+            let last_four_preamble_bytes = self.body_start_search_window(len, buf);
+            self.num_preamble_bytes += len;
+            self.last_four_preamble_bytes = last_four_preamble_bytes;
+            return Err(NetError::UnderflowError(
+                "Not enough bytes to form HTTP preamble".into(),
+            ));
+        }
+
+        let mut cursor = io::Cursor::new(buf);
+
+        let preamble = {
+            let mut rd =
+                BoundReader::from_reader(&mut cursor, HTTP_PREAMBLE_MAX_ENCODED_SIZE as u64);
+            let preamble: StacksHttpPreamble = read_next(&mut rd)?;
+            preamble
+        };
+
+        let preamble_len = cursor.position() as usize;
+        self.set_preamble(&preamble)?;
+
+        Ok((preamble, preamble_len))
+    }
+
+    /// Stream a payload of unknown length.  Only gets called if payload_len() returns None.
+    ///
+    /// Returns Ok((Some((message, num-bytes-consumed)), num-bytes-read)) if we read enough data to
+    /// form a message.  `num-bytes-consumed` is the number of bytes required to parse the message,
+    /// and `num-bytes-read` is the number of bytes read in this call.
+    ///
+    /// Returns Ok((None, num-bytes-read)) if we consumed data (i.e. `num-bytes-read` bytes), but
+    /// did not yet have enough of the message to parse it.  The caller should try again.
+    ///
+    /// Returns Error on irrecoverable error.
+    fn stream_payload<R: Read>(
+        &mut self,
+        preamble: &StacksHttpPreamble,
+        fd: &mut R,
+    ) -> Result<(Option<(StacksHttpMessage, usize)>, usize), NetError> {
+        assert!(self.payload_len(preamble).is_none());
+        match preamble {
+            StacksHttpPreamble::Request(_) => {
+                // HTTP requests can't be chunk-encoded, so this should never be reached
+                unreachable!()
```

Comment:
Similarly to the other comment, I would do an explicit error. However, I see unreachable is used a lot throughout so imagine this is the preference to prevent a plethora of error types?

Reply:
I can go back and convert these to an error.